### ISSUE TITLE
Gridster has a strong tendency to expand downwards instead of left/right while there's ample space there

### ISSUE THREAD

**Issue by GerHobbelt** [ducksboard/gridster.js#96]
How to observe/reproduce: go to the front page and move the blocks to get the top of the layout  like this (digits/letters represent blocks, `_` is empty space):

```
 _ [1][2--2] _ [3]
 _ [4--4][5--5][3]
[6--6][7][8][9][A]
[6--6] _  _ [B] _ 
```

now drag `[1]` on top of block `[2--2]`: it can shift left or move down to get out of the way. Gridster consistently picks the down-move like this:

```
 _  _ [1] _  _ [3]
 _  _ [2--2] _ [3]
 _ [4--4][5--5][A]
[6--6][7][8][9] _ 
[6--6] _  _ [B] _ 
```

thus expanding occupied surface area significantly (4x6 --> 5x6) while there is sufficient space at right, which would keep the surface area as small as possible and makes for a more 'visually sensible' solution:

```
 _  _ [1][2--2][3]
 _ [4--4][5--5][3]
[6--6][7][8][9][A]
[6--6] _  _ [B] _ 
```

This is probably a feature instead of a bug as I assume -- without having had a look at it yet, so this is conjecture --  the culprit here to be the space fitting algorithm used by Gridster.

I was hoping someone else already had constructed a solution for this one...

**Comment by dustmoo** [comment]
The way gridster is written to only expand down regardless of the available side space.

I have added quite a bit of my own functionality to a fork to support side-swaps:

https://github.com/dustmoo/gridster.js

I have further improved the swap algorithm but it is in a branch for a project where I have severely altered gridster code. If the current master doesn't fit your needs let me know and I will see about adapting my new swap algorithm to the master.

**Comment by GerHobbelt** [comment]
Thanks!
I'll have a look in a couple of days when I'm back at the dev console. For me it's either this or using a D3-based annealing space fitter -- as I'm using both D3 and jQuery in this project of mine.
The lure of gridster in my case is when it's 'good looks, ready to go now' as the D3 alternative would certainly provide better visual appeal/feedback ('great looks') but is also 'more work'.

Is that other project you mention open source so I can take a quick peek? I don't mind if it's got a severely modded gridster, I do RTFC anyway.

**Comment by dustmoo** [comment]
I pushed the branch, it is called "mapapp".

Basically, there was a lot of automatic "shifting down" that I commented out, because for my app, I have a static grid and only need widgets to change places.

Basically, I highly modified the  set_player method, added a new method to move the widgets to where I wanted (new_move_widget_to). And created a movement queue. This checks to see if the widgets can fit where the player was and queues them for swaps.

There is also a new, can_placeholder_be_set method, that checks to see if the placeholder can move to the area you are trying to move it to.

I also added support for static widgets, but still requires the static class to be set drag in initialization.

There is a max_rows and max_cols setting, that so far only work if you are not shifting widgets down.

Basically, this is NOT designed for public consumption as my use for it is very specific. But the swap queue works wonderfully.

If you see anything that could be improved, please feel free to contribute. :)

Cheers!

Edit: punctuation, clarity.

**Comment by vonj** [comment]
This ticket is 4 months old.  Are things still the same with (vanilla) gridster?

**Comment by dustmoo** [comment]
Yes, the vanilla gridster doesn't yet support this. My fork works well for this though.

I did not create gridster, and only have recently been given the ability to merge commits into the default repo. Unfortunately I do not have the time available to merge the two at this time. I hope to when I get a break from the big project I am on.
